File: bob/learn/boosting/cpp/LUTTrainer.cpp
#include <bob.learn.boosting/LUTTrainer.h>
#include <bob.learn.boosting/Functions.h>
#include <limits>

bob::learn::boosting::LUTTrainer::LUTTrainer(uint16_t maximumFeatureValue, int numberOfOutputs, SelectionStyle selectionType) :
  m_maximumFeatureValue(maximumFeatureValue),
  m_numberOfOutputs(numberOfOutputs),
  m_selectionType(selectionType),
  _luts(maximumFeatureValue, numberOfOutputs),
  _selectedIndices(numberOfOutputs),
  _gradientHistogram(maximumFeatureValue)
{
}

int32_t bob::learn::boosting::LUTTrainer::bestIndex(const blitz::Array<double,1>& array) const{
  double min = std::numeric_limits<double>::max();
  int32_t minIndex = -1;
  for (int i = 0; i < array.extent(0); ++i){
    if (array(i) < min){
      min = array(i);
      minIndex = i;
    }
  }
  return minIndex;
}

void bob::learn::boosting::LUTTrainer::weightedHistogram(const blitz::Array<uint16_t,1>& features, const blitz::Array<double,1>& weights) const{
  bob::core::array::assertSameShape(features, weights);
  _gradientHistogram = 0.;
  for (int i = features.extent(0); i--;){
    _gradientHistogram((int)features(i)) += weights(i);
  }
}

boost::shared_ptr<bob::learn::boosting::LUTMachine> bob::learn::boosting::LUTTrainer::train(const blitz::Array<uint16_t,2>& trainingFeatures, const blitz::Array<double,2>& lossGradient) const{
  int featureLength = trainingFeatures.extent(1);
  _lossSum.resize(featureLength, m_numberOfOutputs);
  // Compute the sum of the gradient based on the feature values or the loss associated with each feature index
  // Compute the loss for each feature
  for (int featureIndex = featureLength; featureIndex--;){
    for (int outputIndex = m_numberOfOutputs; outputIndex--;){
      weightedHistogram(trainingFeatures(blitz::Range::all(),featureIndex), lossGradient(blitz::Range::all(), outputIndex));
      _lossSum(featureIndex,outputIndex) = - blitz::sum(blitz::abs(_gradientHistogram));
    }
  }

  // Select the most discriminative index (or indices) for classification which minimizes the loss
  //  and compute the sum of gradient for that index
  if (m_selectionType == independent){
    // independent feature selection is used if all the dimension of output use different feature
    // each of the selected feature minimize a dimension of the loss function
    for (int outputIndex = m_numberOfOutputs; outputIndex--;){
      _selectedIndices(outputIndex) = bestIndex(_lossSum(blitz::Range::all(),outputIndex));
    }
  } else {
    // for 'shared' feature selection the loss function is summed over multiple dimensions and
    // the feature that minimized this cumulative loss is used for all the outputs
    blitz::secondIndex j;
    const blitz::Array<double,1> sum(blitz::sum(_lossSum, j));
    _selectedIndices = bestIndex(sum);
  }

  // compute the look-up-tables for the best index
  for (int outputIndex = m_numberOfOutputs; outputIndex--;){
    int selectedIndex = _selectedIndices(outputIndex);
    weightedHistogram(trainingFeatures(blitz::Range::all(), selectedIndex), lossGradient(blitz::Range::all(), outputIndex));

    for (int lutIndex = m_maximumFeatureValue; lutIndex--;){
      _luts(lutIndex, outputIndex) = (_gradientHistogram(lutIndex) > 0) * 2. - 1.;
    }
  }

  // create new weak machine
  return boost::shared_ptr<LUTMachine>(new LUTMachine(_luts.copy(), _selectedIndices.copy()));

}

File: bob/learn/boosting/LUTTrainer.py
from ._library import LUTMachine, weighted_histogram

import numpy

class LUTTrainer():
  """The LUTTrainer contains methods to learn weak trainer using Look-Up-Tables (LUT).
  It can be used for uni-variate and multi-variate binary classification and, as well as for multi-variate regression.
  It requires that the features are discrete and have a maximum value.
  """

  def __init__(self, maximum_feature_value, feature_length, number_of_outputs = 1, selection_type = 'independent'):
    """Initializes the parameters of the LUT Trainer that trains a weak LUTMachine.

    Keyword parameters
      maximum_feature_value (int): The number of entries for the LUT (i.e., the maximum value of the discrete feature)

      feature_length (int): The length of the feature vectors

      number_of_outputs (int): The number of outputs for the multi-variate case.

      selection_type (str):
        The feature selection can be either 'independent' or 'shared'.
        For the independent case the loss function is separately considered for each output.
        For shared selection type the sum of the loss function is taken over the outputs and a single feature is used for all the outputs.
        See Cosmin's thesis for more details.

    """
    self.m_maximum_feature_value = maximum_feature_value
    self.m_feature_length = feature_length
    self.m_number_of_outputs = number_of_outputs
    self.m_selection_type = selection_type

    # pre-allocate arrays for faster access
    self._feature_gradient = numpy.ndarray((self.m_maximum_feature_value, self.m_number_of_outputs))
    self._luts = numpy.ndarray((self.m_maximum_feature_value, self.m_number_of_outputs))
    self._selected_indices = numpy.ndarray((self.m_number_of_outputs,), numpy.int32)
    self._gradient_histogram = numpy.ndarray((self.m_maximum_feature_value,))
    self._loss_sum = numpy.ndarray((self.m_feature_length, self.m_number_of_outputs))


  def train(self, training_features, loss_gradient):

    """Trains a weak LUTMachine.

    The function searches for a features index that minimizes (the length of) the loss gradient and computes the LUT corresponding to that feature index.

    Keyword parameters
      training_features (uint16<#samples, #features>): The training features samples

      loss_gradient (float<#samples, #outputs>): The loss gradient values for the training samples

    Returns
      A (weak) LUTMachine
    """

    # Compute the sum of the gradient based on the feature values or the loss associated with each feature index
    # Compute the loss for each feature
    for feature_index in range(self.m_feature_length):
      for output_index in range(self.m_number_of_outputs):
        weighted_histogram(training_features[:,feature_index], loss_gradient[:,output_index], self._gradient_histogram)
        self._loss_sum[feature_index, output_index] = - numpy.sum(numpy.abs(self._gradient_histogram))

    # Select the most discriminative index (or indices) for classification which minimizes the loss
    #  and compute the sum of gradient for that index
    if self.m_selection_type == 'independent':

      # independent feature selection is used if all the dimension of output use different feature
      # each of the selected feature minimize a dimension of the loss function
      for output_index in range(self.m_number_of_outputs):
        self._selected_indices[output_index] = self._loss_sum[:,output_index].argmin()

    else:  # shared

      # for 'shared' feature selection the loss function is summed over multiple dimensions and
      # the feature that minimized this cumulative loss is used for all the outputs
      accumulated_loss = numpy.sum(self._loss_sum,1)
      self._selected_indices.fill(accumulated_loss.argmin())

    for output_index in range(self.m_number_of_outputs):
      feature_index = self._selected_indices[output_index]
      self._feature_gradient[:,output_index] = weighted_histogram(training_features[:,feature_index], loss_gradient[:,output_index], self.m_maximum_feature_value)

    # Assign the values to LookUp Table
    self._luts.fill(1.)
    self._luts[self._feature_gradient <= 0.0] = -1.

    # create new weak machine
    return LUTMachine(self._luts.copy(), self._selected_indices.copy())

